### PR TITLE
Subtitle system for streaming audo/video

### PR DESCRIPTION
**Pull request by PTwr**
A simple subtitle system to allow for translating games without any riivolution patches or romhacking.

It started as my pet project, R79JAF translation, to allow translation of voice files that had no accompanying on-screen text. 
After boasting about it on discord some translators expressed interest in it so I upgraded it to be more universal.

Target games are ones with a lot of audio/video files streamed on-demand from disc. It won't work on pre-fetched stuff. So stuff like "visual novels" would be prime target.

Basically, it takes json file like this
```
[
  {
    "FileName": "sound/stream/evz136.brstm",
    "Translation": "This time it's a bit of an irregular job.\nKlaus, did you get that thing?",
    "Miliseconds": 5000,
    "Color": "Green",
    "Enabled": true,
    "AllowDuplicate": false,
    "Scale": 2
  }
]
```
And does stuff like this:
R79JAF emulator subtitles
in-battle chatter
cutscene

Under the hood it works same as FileMonitor log,  Simple hook in DVDThread and lookup for filename in dictionary.
As such it can be used for any file read from disc.

Subtitles are (re)loaded when game starts.

Display is done wtih ImGui through OSD, which I upgraded to handle multiple "zones".

Fancy functions:
- file location is same as other stuff from **Load** directory.
- support for multiple subtitle files per game, including nested directories. Quite useful when you have six thousands voice files to translate.
- support for placing subtitles in long-playing audio/video stream through relative offset and timestamp (measured in real time, not emulated)
- html colors
- ugly text size, Dolphin appears to not be prepared for scalable fonts in ImGui.

Possible issues:
- memory management, pointer/reference mixup, and other standard problems C# developer have in C++ :)
- do we want to keep my changes to OSD, or should subtitles be done separately?
- some automatic generator of subtitle file for translator could be useful for non-techie translators, and maybe nicer file format

### CODE REVIEW

File: Source/Core/Subtitles/Helpers.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        else if (parts.size() == 3)
+        {
+          a = 255;
+          r = std::stoul(parts[0], nullptr, 10);
+          g = std::stoul(parts[1], nullptr, 10);
+          b = std::stoul(parts[2], nullptr, 10);
+          return a << 24 | r << 16 | g << 8 | b;
+        }
+      }
+      catch (std::exception x)
```
Comment (BhaaLseN) [inline Source/Core/Subtitles/Helpers.h:62]: We generally don't build with exceptions, so I'm a bit surprised this does something.

Instead, try to use `TryParse` from [`StringUtil.h`](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Common/StringUtil.h#L77).

File: Source/Core/Core/HW/DVD/FileMonitor.cpp [hunk 2]
```diff
@@ -45,6 +45,19 @@ static bool IsSoundFile(const std::string& filename)
...
 }
+// Filtered files
+static bool IsVideoFile(const std::string& filename)
+{
+  std::string extension;
+  SplitPath(filename, nullptr, nullptr, &extension);
+  Common::ToLower(&extension);
+
+  static const std::unordered_set<std::string> extensions = {
+      ".thp",  // 1Wii/Game Cube Video File
```
[the remark this answers is not in the document]
Comment (sepalani) [reply Source/Core/Core/HW/DVD/FileMonitor.cpp:56]: [THP](https://wiki.tockdom.com/wiki/THP_(File_Format)) is a common video format. My only issue with this function is that the extension is hardcoded and can't be changed. It also seems to be used once, what is the point of `IsVideoFile` beside changing the log level?

It'd be nice for this list to be configurable somehow. BTW, there are other existing video formats. Moreover, I'm pretty sure I've seen compressed video files such as `thp.lz` or `thp.lz11` I don't remember exactly. So having a way to make this configurable would be preferred if it has a meaningful use.

[the remark this answers is not in the document]
Comment (BhaaLseN) [reply Source/Core/Core/HW/DVD/FileMonitor.cpp:56]: Oh, derp, it refers to the _video_ file. I was racking my brain about why the "T" from "subTitle" is in the beginning 🤦 
Guess I shouldn't be reviewing code after a long day at work.

**Review by sepalani — CHANGES_REQUESTED** [review]
You'll find below a cursory review. You might want to check the [project coding style](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md) and the [lint bot result](https://dolphin.ci/#/builders/9/builds/2067/steps/3/logs/stdio) to address some of the issues.

I'd really like Dolphin to have a translation/subtitle system/API. However, I'm unsure which form it should take and if your current proposal is the best way to implement one. As you said, you're hooking file loads and it won't work if the file is used later.

> A simple subtitle system to allow for translating games without any riivolution patches or romhacking.

I can understand the rational as it can make translation easier. However, this approach prevents the translation project from being used on real hardware _(which is kinda sad for a translation project)_.

Some games might bundle several texts, voices, videos into a single file. I've rarely seen games using one file (audio, video or text) for a single message which makes your approach very situational.

Other ways that might be used to extract texts or its address/properties of Dolphin (from an external process):
 - Using Dolphin Memory Engine
 - Using File/IO logs and checking Dolphin logs
 - Using log breakpoints and checking Dolphin logs

Possible alternatives to extract texts from Dolphin (to be implemented):
 - Custom HLE hooks
 - Custom breakpoint log messages (to fetch text from specific address/register)
 - Custom File/IO/DVD hooks

For reference, common methods to translate games:
 - Create a game translation patch, using Wiimm ISO Patcher _(based on WIT)_ and/or Riivolution.
 - IIRC, libretro can _(try to)_ translate using a screenshot, send it to an OCR and send its results to a translator.
 - Text extracting tool can be used _(Textractor, for instance)_ and its results can be sent to a translator.

File: Source/Core/Core/HW/DVD/FileMonitor.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PTwr) [reply Source/Core/Core/HW/DVD/FileMonitor.cpp:56]: @sepalani the idea was to add more video extensions here.
Small problem with FileMonitor is that it (ab)uses log levels to make sound logs blue.
I added video color in same way, but reusing bad idea is still a bad idea.
I think filetype/coloring in FileMonitor has to be reworked to more civilized form.

File: Source/Core/DolphinLib.vcxproj [hunk 3]
```diff
@@ -69,31 +69,33 @@
...
+  <ItemGroup>
+    <ClInclude Include="Subtitles\Helpers.h" />
+    <ClInclude Include="Subtitles\WebColors.h" />
+    <ClInclude Include="Subtitles\TranslationEntry.h" />
+    <ClInclude Include="Subtitles\Subtitles.h" />
+    <ClInclude Include="VideoCommon\OnScreenDisplay.enum.h" />
+  </ItemGroup>
+  <ItemGroup>
+    <ClCompile Include="Subtitles\Subtitles.cpp" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (PTwr) [reply Source/Core/DolphinLib.vcxproj]: Oh, that's just something Visual Studio made. I'll get rid of it.

File: Source/Core/Subtitles/Helpers.h (hunk 1 above)
Comment (PTwr) [reply Source/Core/Subtitles/Helpers.h:68]: I think I'll be getting rid of that part. Its from before I added support for webcolors.
I do not think we need anything other than named webcolors and hex notation anyway.

File: Source/Core/VideoCommon/OnScreenDisplay.enum.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (PTwr) [reply Source/Core/VideoCommon/OnScreenDisplay.enum.h]: I removed it already. It's left over from when I tried to make subtitles as a dll plugin.

**Comment by PTwr** [comment]
> You'll find below a cursory review. You might want to check the [project coding style](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md) and the [lint bot result](https://dolphin.ci/#/builders/9/builds/2067/steps/3/logs/stdio) to address some of the issues.
> 
> I'd really like Dolphin to have a translation/subtitle system/API. However, I'm unsure which form it should take and if your current proposal is the best way to implement one. As you said, you're hooking file loads and it won't work if the file is used later.
> 
> > A simple subtitle system to allow for translating games without any riivolution patches or romhacking.
> 
> I can understand the rational as it can make translation easier. However, this approach prevents the translation project from being used on real hardware _(which is kinda sad for a translation project)_.
> 
> Some games might bundle several texts, voices, videos into a single file. I've rarely seen games using one file (audio, video or text) for a single message which makes your approach very situational.
> 
> Other ways that might be used to extract texts or its address/properties of Dolphin (from an external process):
> 
>     * Using Dolphin Memory Engine
> 
>     * Using File/IO logs and checking Dolphin logs
> 
>     * Using log breakpoints and checking Dolphin logs
> 
> 
> Possible alternatives to extract texts from Dolphin (to be implemented):
> 
>     * Custom HLE hooks
> 
>     * Custom breakpoint log messages (to fetch text from specific address/register)
> 
>     * Custom File/IO/DVD hooks
> 
> 
> For reference, common methods to translate games:
> 
>     * Create a game translation patch, using Wiimm ISO Patcher _(based on WIT)_ and/or Riivolution.
> 
>     * IIRC, libretro can _(try to)_ translate using a screenshot, send it to an OCR and send its results to a translator.
> 
>     * Text extracting tool can be used _(Textractor, for instance)_ and its results can be sent to a translator.

This little fork came as a result of my [translation effort of R79JAF](https://github.com/PTwr/0079tl), as a companion to riivolution patch with text and textures.
I might have picked wrong screenshot to explain this situation, but game has 6 **thousands** voice files, played all around the game, including in-battle, without any accompanying text field on screen.

In example I provided at start I do not use subtitles, as I am in process of patching all on-screen text.
Subtitles are using in places like this:
cutscene
or like this:
chatter during battle

**Comment by Anuskuss** [comment]
I've been waiting for something like this. Great job!
Although I was wondering, could this be expanded to also support swapping tracks with your own audio files, preferrably with audio formats that the Wii can't even play (e.g. 48kHz FLAC)?

File: Source/Core/Subtitles/Helpers.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+{
+  OSD::AddMessage(err, 2000, OSD::Color::RED);
+  ERROR_LOG_FMT(SUBTITLES, "{}", err);
+}
+
+u32 TryParsecolor(picojson::value raw, u32 defaultColor)
+{
+  if (raw.is<double>())
+    return raw.get<double>();
+  else
```
[the remark this answers is not in the document]
Comment (PTwr) [reply Source/Core/Subtitles/Helpers.h]: I'm all in for more syntax than needed :)
Before PR I trimmed bunch of curlies on one-liners because I saw such style all around the code, I'll have to do those lint style checks.

File: Source/Core/Subtitles/Helpers.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+          r = std::stoul(parts[0], nullptr, 10);
+          g = std::stoul(parts[1], nullptr, 10);
+          b = std::stoul(parts[2], nullptr, 10);
+          return a << 24 | r << 16 | g << 8 | b;
+        }
+      }
+      catch (std::exception x)
+      {
+        Error("Invalid color: " + str);
+      }
```
[the remark this answers is not in the document]
Comment (PTwr) [reply Source/Core/Subtitles/Helpers.h]: this abomination has been removed

File: Source/Core/Core/HW/DVD/FileMonitor.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PTwr) [reply Source/Core/Core/HW/DVD/FileMonitor.cpp]: @sepalani I dipped into logger code, and it appears we'd have to churn through it top-to-bottom to add custom colors per log. Currently FileMonitor abuses log level to make sound logs blue.
Maybe we should just replace IsSoundFile with IsSoundOrVideFile?

File: Source/Core/Subtitles/Helpers.h (hunk 1 above)
Comment (PTwr) [reply Source/Core/Subtitles/Helpers.h]: hex parsing replaced with tryparse, rest removed.

File: Source/Core/Core/HW/DVD/DVDThread.cpp [hunk 7]
```diff
@@ -343,6 +344,7 @@ void DVDThread::DVDThreadMain()
     while (m_request_queue.Pop(request))
     {
       m_file_logger.Log(*m_disc, request.partition, request.dvd_offset);
+      Subtitles::OnFileAccess(*m_disc, request.partition, request.dvd_offset);
```
[the remark this answers is not in the document]
Comment (PTwr) [reply Source/Core/Core/HW/DVD/DVDThread.cpp:347]: Subtitles are disabled until game with any is loaded.
OnFileAccess has a wall of early exits in similar style to FileMonitor.

If both FileMonitor and Subtitles are running, there will be duplication in stuff like ```file_system->FindFileInfo(offset)```, but unifying both monitors would make a messy code.

```c++
void OnFileAccess(const DiscIO::Volume& volume, const DiscIO::Partition& partition, u64 offset)
{
  if (!_subtitlesInitialized)
    return;
  ...
```

**Comment by PTwr** [comment]
> I've been waiting for something like this. Great job! Although I was wondering, could this be expanded to also support swapping tracks with your own audio files, preferrably with audio formats that the Wii can't even play (e.g. 48kHz FLAC)?

TLDR: Yes, but...

There was a discuss about something like that on modding Discord.  Idea being that if we can't really make emulated (or real) Wii to play super duper 4k video, we can just make Dolphin play it.
Basically, if we can monitor IO to streaming file, we can just play custom audio/video in popup for as long as game is streaming original file.

There are some issues to be resolved to get such system to work
- getting some good audio/video player that can be controlled programmatically
- syncing real and emulated time or frames
- mixing replaced sound with background music
- stopping playback when player skips it

Vdeo:
Simplest way for video replacement, would be to just pause the game, play the video until player hits **skip**, resume game, send virtual **skip** keypress to game. No muxing issues, no time sync problems. 
You could monitor for continuos reads and sync custom video with real... but if its full screen video, why bother.
It would be quite easy to do, and rather problem free. But it will only work for cutscenes in exclusive playback.

However, there are more complicated video playbacks. Some are looped, some are mixed into gameplay scene.
TLDR: Cutscenes/intro/outro yes, anything more complex no.

Sound:
Unlike video, sounds are bound to be played multiple at a time, so just muting game won't work, and being played in-game makes them prone to desync, and unsuitable for pause+overlay trick.
It's still doable, but you'd need to somehow mute replaced file, simplest way would be to provide silent stream of same length.
Issue here would be prefetched sounds. in R79JAF, and some other voice-heavy games I looked at, voices and background music are streamed directly from disc, but sound effects are pre-fetched thus undetectable for FileMonitor.

**Comment by Impeta** [comment]
@PTwr @BhaaLseN @sepalani Hi, sorry for bothering but will this be ever merged? I'm asking because there's this Tetra Trackers ENG translation project, and it's using this PR for showing up subtitles on Tetra/other narrating characters speaking amidst games. Right now it's narrowing in soon for a new update.

https://youtu.be/OT1vQwp1ZJE?t=7482
https://gamebanana.com/mods/413831

snapshot

**Comment by sepalani** [comment]
This seems quite unlikely unless this 3yo PR is rebased and some of the issues are addressed (e.g. [project coding style](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md), passing buildbot tests, ...). I'll be more than happy to review this PR again when it's updated.

**Comment by PTwr** [comment]
@Impeta I had to take a break from being alive for... uh... 3 years?! Damn. I am still in process of picking up pieces, but I'll should be able to find time to dust this off soon-ish.

@sepalani this MR is garbage, because I first tried to offload subtitles to DLL. Then I discovered C++ Name Mangling, and after failing to build time machine to correct that atrocity I quickly chopped it into built-in mechanism. 

If I remember correctly there's bunch of issues to fix, code quality is least of them:
- File format I cobbled up is garbage, because its the case of tweaking it until it worked despite limitations of system.
- If I remember correctly, I had to include some single-file parser for... JSON? Can't remember details... but it should probably be changed to use something that's already in repo to minimize changes. Dolphin is using various config files, so there should be something handy.
- OSD handling was done horribly, as it was done wile learning OSD system.
- text rendering sucked because fonts are ridiculously complicated nowadays
- Something about adding fields to structs and automatic copy-ctors ruining my mental health. I remember almost biting my keyboard when trying to get it to work, because C++ compiler in its traditional fashion showed errors in random code templates rather than actual code.

It should be split in parts:
- file offset logging, which is **very** useful when reverse-engineering custom file formats. And pretty much only way to do non-invasive subtitling anyway.
- extended OSD "sections", or maybe switch to "window", to not mangle current OSD with subtitles.
- subtitles themselves.

I would very much like to somehow keep it separate from Dolphin, as such subtitle system is bound to need some game-specific logic.
Sadly "native" plugins are no go, as between name-mangling and cross-platform it would be easier to just add scripting engine instead.

If there was option to add some basic scripting, subtitles could be a plugin instead of built-in functions. Thus easy to tweak for specific game (eg. timestamp vs file offset).

(Of course there's the issue of not ruining framerate with script-plugins. If I remember correctly, this crappy built-in had negligible performance cost. Scripts would be more annoying to keep fast.)
